syntax = "proto3";

import "CaseInfo.proto";
import "Empty.proto";

package rips;

service Commands
{
	rpc Execute(CommandParams) returns(CommandReply) {}
}

message FilePathRequest
{
	string path = 1;
}

message ReplaceCaseRequest
{
	string newGridFile = 1;
	int32  caseId      = 2;
}

message ReplaceSourceCasesRequest
{
	string gridListFile = 1;
}

message ExportMultiCaseRequest
{
	string gridListFile = 1;
}

enum SnapshotType
{
	VIEWS = 0;
	PLOTS = 1;
	ALL   = 2;
}

message ExportSnapshotsRequest
{
	SnapshotType type   = 1;
	string       prefix = 2;
}

message ExportPropertyRequest
{
	int32 caseId          = 1;
	int32 timeStep        = 2;
	string property       = 3;
	string eclipseKeyword = 4;
	double undefinedValue = 5;
	string exportFile     = 6;
}

message ExportPropertyInViewsRequest
{
	int32 caseId              = 1;
	repeated string viewNames = 2;
	double undefinedValue     = 3;
}

enum CompdatExportSplit
{
	UNIFIED_FILE                      = 0;
	SPLIT_ON_WELL                     = 1;
	SPLIT_ON_WELL_AND_COMPLETION_TYPE = 2;
}

enum CompdatExportType
{
	TRANSMISSIBILITIES                     = 0;
	WPIMULT_AND_DEFAULT_CONNECTION_FACTORS = 1;
	NO_COMPLETIONS                         = 2;
}

enum CompdatCombinationMode
{
	INDIVIDUALLY = 0;
	COMBINED     = 1;
}

message ExportWellPathCompRequest 
{
	int32 caseId                           = 1;
	int32 timeStep                         = 2;
	repeated string wellPathNames          = 3;
	CompdatExportSplit fileSplit           = 4;
	CompdatExportType compdatExport        = 5;
	bool includePerforations               = 6;
	bool includeFishbones                  = 7;
	bool excludeMainBoreForFishbones       = 8;
	CompdatCombinationMode combinationMode = 9;
}


message SetTimeStepParams
{
	int32 caseId   = 1;
	int32 timeStep = 2;
}

message SetMainWindowSizeParams
{
	int32 height = 1;
	int32 width  = 2;
}


message CommandParams
{
	oneof params
	{
		FilePathRequest              openProject                      = 1;
		Empty                        closeProject                     = 2;
		FilePathRequest              setStartDir                      = 3;
		FilePathRequest              loadCase                         = 4;
		ReplaceCaseRequest           replaceCase                      = 5;
		ReplaceSourceCasesRequest    replaceSourceCases               = 6;
		ExportMultiCaseRequest       exportMultiCaseSnapshots         = 7;
		ExportSnapshotsRequest       exportSnapshots                  = 8;
		ExportPropertyRequest        exportProperty                   = 9;
		ExportPropertyInViewsRequest exportPropertyInViews            = 10;
		ExportWellPathCompRequest    exportWellPathCompletions        = 11;
		//ExportSimWellPathFracRequest exportSimWellFractureCompletions = 12;
		//ExportMswRequest             exportMsw                        = 13;
		//ExportWellPathRequest        exportWellPaths                  = 14;
		//ExportVisibleCellsRequest    exportVisibleCells               = 15;
		//SetExportFolderRequest       setExportFolder                  = 16;
		//RunOctaveScriptRequest       runOctaveScript                  = 17;
		SetMainWindowSizeParams      setMainWindowSize                = 18;
		//ComputeCaseGroupStatRequest  computeCaseGroupStatistics       = 19;
		SetTimeStepParams            setTimeStep                      = 20;
		//ScaleFractureTemplateRequest scaleFractureTemplate            = 21;
		//SetFracContainmentRequest    setFractureContainment           = 22;
		//CreateMultipleFracRequest    createMultipleFractures          = 23;
		//CreateLgrForCompRequest      createLgrForCompletions          = 24;
		//CreateSatPressPlotRequest    createSaturationPressurePlots    = 25;		
	}
}

message CommandReply
{
	oneof result
	{
		Empty     emptyResult = 1;
		Case   loadCaseResult = 2;
	}
}

